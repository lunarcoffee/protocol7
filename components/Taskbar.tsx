'use client';

import { useProcessInfo } from './contexts/ProcessInfoContext';
import { LauncherButton } from './LauncherButton';

export const Taskbar = () => {
  const [processInfo, updateProcessInfo] = useProcessInfo();

  return (
    <div className="absolute bottom-0 w-full h-10 pl-4 bg-gradient-to-t from-aero-tint-dark/50 via-95% via-aero-tint/40 to-white/40 backdrop-blur-xs border-t border-t-black flex flex-row items-center">
      <div className="flex left-4 -mt-1">
        <LauncherButton />
      </div>
    </div>
  );
};
