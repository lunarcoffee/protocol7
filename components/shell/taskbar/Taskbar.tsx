'use client';

import Image from 'next/image';

import { useWindowManager } from '@/hooks/windows/useWindowManager';
import NetworkIcon from '@/public/icons/network.svg';
import VolumeHighIcon from '@/public/icons/volume-high.svg';

import { Clock } from './Clock';
import { LauncherButton } from './LauncherButton';
import { SystemTray } from './SystemTray';
import { WindowList } from './WindowList';

export const Taskbar = () => {
  const [{ windows }] = useWindowManager();
  const windowArray = Array.from(windows.values());

  return (
    <div
      className={`
        absolute bottom-0 z-[calc(infinity)] flex h-10 w-full max-w-full
        flex-row items-center border-t border-t-aero-tint-darkest/85
        bg-gradient-to-t from-aero-tint-dark/80 to-aero-tint-dark/70
        backdrop-blur-xs
      `}
    >
      {/* white top highlight */}
      <div
        className={`
          absolute top-0 left-0 z-20 h-0.5 w-full bg-gradient-to-b from-white/30
        `}
      />
      {/* darkened background over launcher button */}
      <div
        className={`
          flex h-full items-center bg-gradient-to-r from-aero-tint-darkest/15
          via-aero-tint-darkest/20 via-[calc(100%-1.5rem] pr-5 pl-4
        `}
      >
        <div className="left-4 z-30 -mt-1 mr-4">
          <LauncherButton />
        </div>
      </div>
      <div
        className={`
          z-10 -mx-5 flex h-full min-w-0 shrink grow flex-row items-center
          gap-1.5
        `}
      >
        <WindowList windows={windowArray} />
      </div>
      {/* darkened background over tray and clock */}
      <div
        className={`
          flex h-full flex-row items-center bg-gradient-to-l
          from-aero-tint-darkest/15 via-aero-tint-darkest/20
          via-[calc(100%-1.5rem)] pr-2 pl-5
        `}
      >
        <div className="ml-3 shrink-0">
          {/* TODO: eventually this module should be attached to a global network manager state thing */}
          <SystemTray
            items={[
              {
                renderIcon: () => (
                  <Image
                    src={VolumeHighIcon}
                    alt="network icon"
                    className="w-7 p-[0.27rem]"
                  />
                ),
                renderPane: () => <p></p>,
              },
              {
                renderIcon: () => (
                  <Image
                    src={NetworkIcon}
                    alt="network icon"
                    className="w-7 p-[0.4rem]"
                  />
                ),
                renderPane: () => <p></p>,
              },
            ]}
          />
        </div>
        <div className="z-30 ml-1 h-full shrink-0">
          <Clock />
        </div>
      </div>
    </div>
  );
};
