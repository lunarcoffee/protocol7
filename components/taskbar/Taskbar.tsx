'use client';

import { useProcessInfo } from '../contexts/ProcessInfoContext';
import { Clock } from './Clock';
import { LauncherButton } from './LauncherButton';
import { useToggle } from '@/hooks/useToggle';
import { SystemTray } from './SystemTray';

import NetworkIcon from '../../public/icons/network.svg';
import WirelessIcon from '../../public/icons/wireless.svg';
import BatteryIcon from '../../public/icons/battery.svg';
import VolumeHighIcon from '../../public/icons/volume-high.svg';
import Image from 'next/image';

export const Taskbar = () => {
  const [processInfo, updateProcessInfo] = useProcessInfo();
  const [isActive, toggleActive] = useToggle();

  return (
    <div className="absolute bottom-0 w-full h-10 pl-4 pr-2 bg-gradient-to-t from-aero-tint-dark/50 via-95% via-aero-tint/40 to-white/40 backdrop-blur-xs border-t border-t-aero-tint-darkest/85 flex flex-row items-center">
      <div className="left-4 -mt-1">
        <LauncherButton active={isActive} onClick={toggleActive} />
      </div>
      <div className="grow"></div>
      <div className="">
        {/* TODO: eventually this module should be attached to a global network manager state thing */}
        <SystemTray
          items={[
            {
              renderIcon: () => (
                <Image
                  src={VolumeHighIcon}
                  alt="network icon"
                  className="w-7 p-[0.25rem]"
                />
              ),
              renderPane: () => <p></p>,
            },
            {
              renderIcon: () => (
                <Image
                  src={NetworkIcon}
                  alt="network icon"
                  className="w-7 p-[0.4rem]"
                />
              ),
              renderPane: () => <p></p>,
            },
            {
              renderIcon: () => (
                <Image
                  src={WirelessIcon}
                  alt="network icon"
                  className="w-7 p-[0.4rem]"
                />
              ),
              renderPane: () => <p></p>,
            },
          ]}
        />
      </div>
      <div className="h-full ml-1">
        <Clock />
      </div>
    </div>
  );
};
