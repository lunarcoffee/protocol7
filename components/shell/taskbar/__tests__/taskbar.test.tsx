import { DateTime } from 'luxon';

import { render } from '@/utils/tests/renderSystemContext';

import { Taskbar } from '../Taskbar';

jest.mock('/hooks/useCurrentTime', () => ({
  // should be rendered the same in any locale
  useCurrentTime: () =>
    DateTime.fromFormat('08/17/2010', 'MM/dd/yyyy') as DateTime<true>,
}));

describe('taskbar', () => {
  it('renders correctly with no windows', () => {
    const { asFragment } = render(<Taskbar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
