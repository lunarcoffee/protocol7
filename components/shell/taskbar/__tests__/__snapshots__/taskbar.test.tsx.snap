// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`taskbar renders correctly with no windows 1`] = `
<DocumentFragment>
  <div
    class="
        absolute bottom-0 z-[calc(infinity)] flex h-10 w-full max-w-full
        flex-row items-center border-t border-t-aero-tint-darkest/85
        bg-gradient-to-t from-aero-tint-dark/80 to-aero-tint-dark/70
        backdrop-blur-xs
      "
  >
    <div
      class="
          absolute top-0 left-0 z-20 h-0.5 w-full bg-gradient-to-b from-white/30
        "
    />
    <div
      class="
          flex h-full items-center bg-gradient-to-r from-aero-tint-darkest/15
          via-aero-tint-darkest/20 via-[calc(100%-1.5rem] pr-5 pl-4
        "
    >
      <div
        class="left-4 z-30 -mt-1 mr-4"
      >
        <div
          class="group z-0 flex size-12 cursor-pointer flex-row justify-center rounded-full bg-aero-tint-dark shadow-[0_0_0.3rem] ring shadow-white/60 ring-aero-tint-darkest transition duration-100 hover:shadow-[0_0_0.5rem] hover:shadow-white/70"
        >
          <div
            class="
        absolute z-10 h-7 w-12 rounded-t-3xl rounded-b-sm bg-gradient-to-b
        from-white/40 to-white/5 inset-shadow-sm inset-shadow-white/40
        transition duration-100
        group-hover:from-white/50 group-hover:inset-shadow-white/60
      "
          />
          <div
            class="
        absolute z-20 mt-[1.62rem] h-2 w-10 rounded-[50%] bg-gradient-to-b
        from-aero-tint-dark from-40% to-transparent to-40%
      "
          />
          <div
            class="
        absolute z-30 size-12 rounded-full bg-radial-[at_50%_100%]
        from-aero-tint to-transparent to-50% transition duration-100
        group-hover:to-60%
      "
          />
          <img
            alt="launcher icon"
            class="
        absolute z-30 mt-1 size-10 opacity-70 drop-shadow-[0_0_0]
        drop-shadow-transparent transition duration-100
        group-hover:opacity-100 group-hover:drop-shadow-[0_0_2px]
        group-hover:drop-shadow-white/50
      "
            data-nimg="1"
            decoding="async"
            draggable="false"
            height="40"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
        </div>
      </div>
    </div>
    <div
      class="
          z-10 -mx-5 flex h-full min-w-0 shrink grow flex-row items-center
          gap-1.5
        "
    />
    <div
      class="
          flex h-full flex-row items-center bg-gradient-to-l
          from-aero-tint-darkest/15 via-aero-tint-darkest/20
          via-[calc(100%-1.5rem)] pr-2 pl-5
        "
    >
      <div
        class="ml-3 shrink-0"
      >
        <div
          class="flex flex-row justify-center"
        >
          <div
            class="
      group relative h-full rounded-xs ring-aero-tint-darkest/80 transition
      duration-100
      hover:ring hover:inset-shadow-[0_2px_6px] hover:inset-shadow-white/30
    "
          >
            <div
              class="drop-shadow-xs drop-shadow-aero-tint-darkest"
            >
              <img
                alt="network icon"
                class="w-7 p-[0.27rem]"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
          </div>
          <div
            class="
      group relative h-full rounded-xs ring-aero-tint-darkest/80 transition
      duration-100
      hover:ring hover:inset-shadow-[0_2px_6px] hover:inset-shadow-white/30
    "
          >
            <div
              class="drop-shadow-xs drop-shadow-aero-tint-darkest"
            >
              <img
                alt="network icon"
                class="w-7 p-[0.4rem]"
                data-nimg="1"
                decoding="async"
                height="40"
                loading="lazy"
                src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
                srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
                style="color: transparent;"
                width="40"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="z-30 ml-1 h-full shrink-0"
      >
        <div
          class="
      group relative h-full rounded-xs ring-aero-tint-darkest/80 transition
      duration-100
      hover:ring hover:inset-shadow-[0_2px_6px] hover:inset-shadow-white/30
    "
        >
          <div
            class="
          absolute h-full w-full bg-radial-[at_50%_140%] from-transparent
          to-transparent to-70% transition duration-100
          group-hover:from-aero-tint/80
        "
          />
          <div
            class="
          flex h-full flex-col items-center justify-center px-2 text-gray-200
          text-shadow-aero-tint-darkest/50 text-shadow-md
        "
          >
            <p
              class="z-10 text-xs"
            >
              12:00 AM
            </p>
            <p
              class="z-10 text-xs"
            >
              8/17/2010
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
